plugins/auth-backend-module-bitbucket-server-provider/src/helpers.ts

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import fetch from 'node-fetch';
+import { PassportProfile } from '@backstage/plugin-auth-node';
+
+export async function fetchProfile(options: {
+  host: string;
+  accessToken: string;
+}): Promise<PassportProfile> {
+  const { host, accessToken } = options;
+  // Get current user name
+  let whoAmIResponse;
+  try {
+    whoAmIResponse = await fetch(
+      `https://${host}/plugins/servlet/applinks/whoami`,
+      {
+        headers: {
+          Authorization: `Bearer ${accessToken}`,
+        },
+      },
+    );
+  } catch (e) {
+    throw new Error(`Failed to retrieve the username of the logged in user`);
+  }
+
+  // A response.ok check here would be worthless as the Bitbucket API always returns 200 OK for this call
+  const username = whoAmIResponse.headers.get('X-Ausername');
```

Comment:
The `X-Ausername` header may not always conform to the value expected by the Bitbucket Server REST API to designate the user. You may have to massage the data here to convert `@` to `_`.

![bitbucket-server-whoami-headers](https://github.com/user-attachments/assets/7e8bdca9-84b0-46f8-a3a0-a32b7ca3bbc3)

Organizations that populate their users in Bitbucket using SAML/LDAP providers often designate usernames via email addresses. However, the Bitbucket username API does not accept usernames with `@` symbols. As such, while the display name may contain the symbol, the REST API requires that `@` symbols be replaced with `_`. 

For example, a user with the username of `ddarji@company.com` would have the username of `ddarji_company.com.` 

![bitbucket-server-user-response](https://github.com/user-attachments/assets/c8bd525a-3934-4f91-afe5-c12a75983571)
![bitbucket-server-user-headers](https://github.com/user-attachments/assets/cb55acb2-a58f-4520-b623-51ce413d4ed7)

Reply:
That's a good catch, thanks for pointing out 🙌🏻 . Do you think this transformation should always be done or it could be optionally enabled via config? As I mentioned in the comment above, I'm just porting the legacy provider code to the new auth module format, I wonder if it would be a good idea to make the changes you suggested in a follow-up pull request. Let's see what @freben says 🙏🏻

Reply:
I believe the conversion should always be done, but as you said, it can be done in a separate PR. If someone really needs it, they will propose the change. 

Normally, that someone would have been me, but my organization is no longer using this provider, so there's no reason to hold this back. 😅